src/hotspot/share/classfile/stackMapTable.cpp

```diff
@@ -29,38 +29,50 @@
 #include "oops/oop.inline.hpp"
 #include "runtime/handles.inline.hpp"
 
-StackMapTable::StackMapTable(StackMapReader* reader, StackMapFrame* init_frame,
-                             u2 max_locals, u2 max_stack,
-                             char* code_data, int code_len, TRAPS) {
-  _code_length = code_len;
+StackMapTable::StackMapTable(StackMapReader* reader, TRAPS) {
+  _code_length = reader->code_length();
   _frame_count = reader->get_frame_count();
   if (_frame_count > 0) {
-    _frame_array = NEW_RESOURCE_ARRAY_IN_THREAD(THREAD,
-                                                StackMapFrame*, _frame_count);
-    StackMapFrame* pre_frame = init_frame;
-    for (int32_t i = 0; i < _frame_count; i++) {
-      StackMapFrame* frame = reader->next(
-        pre_frame, i == 0, max_locals, max_stack,
-        CHECK_VERIFY(pre_frame->verifier()));
-      _frame_array[i] = frame;
-      int offset = frame->offset();
-      if (offset >= code_len || code_data[offset] == 0) {
-        frame->verifier()->verify_error(
-            ErrorContext::bad_stackmap(i, frame),
-            "StackMapTable error: bad offset");
-        return;
+    _frame_array = new GrowableArray<StackMapFrame*>();
+    while (!reader->at_end()) {
+      StackMapFrame* frame = reader->next(CHECK_VERIFY(reader->prev_frame()->verifier()));
+      if (frame != nullptr) {
+        _frame_array->push(frame);
       }
-      pre_frame = frame;
     }
+    reader->check_end(CHECK);
+    // Correct frame count based on how many actual frames are generated
+    _frame_count = _frame_array->length();
+  }
+}
+
+void StackMapReader::check_offset(StackMapFrame* frame, TRAPS) {
+  int offset = frame->offset();
+  if (offset >= _code_length || _code_data[offset] == 0) {
+    _verifier->verify_error(
+        ErrorContext::bad_stackmap(0, frame),
+        "StackMapTable error: bad offset");
+  }
+}
+
+void StackMapReader::check_size(TRAPS) {
+  if (_frame_count < _parsed_frame_count) {
+    StackMapStream::stackmap_format_error("wrong attribute size", CHECK);
+  }
+}
+
+void StackMapReader::check_end(TRAPS) {
+  assert(_stream->at_end(), "must be");
+  if (_frame_count != _parsed_frame_count) {
+    StackMapStream::stackmap_format_error("wrong attribute size", CHECK);
```

Comment:
```suggestion
    StackMapStream::stackmap_format_error("wrong attribute size", THREAD);
```
The method will return no matter what.